components/de_amp/browser/de_amp_url_loader.cc

```diff
@@ -62,54 +64,72 @@ DeAmpURLLoader::~DeAmpURLLoader() = default;
 
 void DeAmpURLLoader::OnBodyReadable(MojoResult) {
   if (state_ == State::kSending) {
-    // The pipe becoming readable when kSending means all buffered body has
-    // already been sent.
-    ForwardBodyToClient();
+    if (bytes_remaining_in_buffer_ > 0) {
+      SendBufferedBodyToClient();
+    } else {
+      ForwardBodyToClient();
+    }
     return;
   }
   if (!CheckBufferedBody(kReadBufferSize)) {
     return;
   }
-
-  if (!MaybeRedirectToCanonicalLink()) {
+  bool redirected = MaybeRedirectToCanonicalLink();
```

Comment:
Does it mean that now we usually do 3x work to check the body parts againt the regexs?

Reply:
Not in the normal case - only if it's an AMP page.